pkg/dns/aws/dns.go

```diff
@@ -0,0 +1,248 @@
+	if err != nil {
+		return fmt.Errorf("failed to list hosted zones: %v", err)
+	}
+
+	m.zones = zones
+	return nil
+}
+
+// findClusterPublicZone finds the public zone given the base domain.
+func (m *Manager) findClusterPublicZone() *route53.HostedZone {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	for _, zwt := range m.zones {
+		if aws.StringValue(zwt.Zone.Name) == m.config.BaseDomain && !aws.BoolValue(zwt.Zone.Config.PrivateZone) {
+			return zwt.Zone
+		}
+	}
+	return nil
+}
+
+// findClusterPrivateZone finds the private zone given the base domain and
+// cluster ID.
+func (m *Manager) findClusterPrivateZone() *route53.HostedZone {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	for _, zwt := range m.zones {
+		if aws.StringValue(zwt.Zone.Name) == m.config.BaseDomain && aws.BoolValue(zwt.Zone.Config.PrivateZone) {
+			for _, tag := range zwt.Tags {
+				if aws.StringValue(tag.Key) == "tectonicClusterID" && aws.StringValue(tag.Value) == m.config.ClusterID {
+					return zwt.Zone
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// EnsureAlias will create an A record for domain/target in both the public and
+// private zones. The target must correspond to the hostname of an ELB which
+// will be automatically discovered.
+func (m *Manager) EnsureAlias(domain, target string) error {
+	if len(domain) == 0 {
+		return fmt.Errorf("domain is required")
+	}
+	if len(target) == 0 {
+		return fmt.Errorf("target is required")
+	}
+
+	// Ensure the zone cache is up to date.
+	err := m.updateZoneCache()
+	if err != nil {
+		return fmt.Errorf("failed to update hosted zone cache: %v", err)
+	}
+
+	publicZone := m.findClusterPublicZone()
+	privateZone := m.findClusterPrivateZone()
+	if publicZone == nil {
+		return fmt.Errorf("failed to find public zone for cluster")
+	}
+	if privateZone == nil {
+		return fmt.Errorf("failed to find private zone for cluster")
+	}
+
+	// Find the target hosted zone of the load balancer attached to the service.
+	// TODO: cache it?
+	var targetHostedZoneID string
+	loadBalancers, err := m.ELB.DescribeLoadBalancers(&elb.DescribeLoadBalancersInput{})
+	if err != nil {
+		return fmt.Errorf("failed to describe load balancers: %v", err)
+	}
+	for _, lb := range loadBalancers.LoadBalancerDescriptions {
+		logrus.Debugf("found load balancer: %s (DNS name: %s, zone: %s)", aws.StringValue(lb.LoadBalancerName), aws.StringValue(lb.DNSName), aws.StringValue(lb.CanonicalHostedZoneNameID))
+		if aws.StringValue(lb.CanonicalHostedZoneName) == target {
+			targetHostedZoneID = aws.StringValue(lb.CanonicalHostedZoneNameID)
+			break
+		}
+	}
+	if len(targetHostedZoneID) == 0 {
+		return fmt.Errorf("couldn't find hosted zone ID of target ELB with domain name %s", target)
+	}
+
+	// Create or update an alias from the wildcard domain to the service load
+	// balancer hostname in both zones.
+	// TODO: only make this call if the records don't exist.
+	updateRecord := func(zoneID string) error {
```

Comment:
This is the part that external-dns can do for you right now. Btw, if I set the --registry=noop flag, I don't have it create the ugly TXT records.